{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'The total branch length of the first tree is 0.'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import tskit\n",
    "\n",
    "def calculate_branch_length(file_path):\n",
    "    # Load the tree sequence from the given file path\n",
    "    ts = tskit.load(file_path)\n",
    "    # Get the first tree in the tree sequence\n",
    "    first_tree = ts.first()\n",
    "    # Calculate the total branch length of the first tree\n",
    "    total_branch_length = sum(first_tree.branch_length(u) for u in first_tree.nodes())\n",
    "    return f'The total branch length of the first tree is {total_branch_length}.' \n",
    "\n",
    "calculate_branch_length()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts = tskit.load('./data/sample.trees')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "f = ts.first()\n",
    "for i in f.nodes():\n",
    "    print(i)\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from graph import generate\n",
    "from langchain_core.messages import HumanMessage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "user_input = \"how many sites does second tree has in the given treesequence\"\n",
    "user_input = [HumanMessage(content=user_input)]\n",
    "\n",
    "graph_state = {\"messages\": user_input, \"iterations\": 0, \"error\": \"\", \"input_files\": \"./data/sample.trees\", \"next\": None, \"generation\": None, \"result\": None}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[HumanMessage(content='how many sites does second tree has in the given treesequence', additional_kwargs={}, response_metadata={})]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "user_input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-- Generating Code -- \n",
      "```{currentmodule} tskit\n",
      "```\n",
      "\n",
      "(sec_tskit_getting_started)=\n",
      "# Getting started with {program}`tskit`\n",
      "\n",
      "You've run some simulations or inference methods, and you now have a \n",
      "{class}`TreeSequence` object; what now? This tutorial is aimed \n",
      "users who are new to {program}`tskit` and would like to get some\n",
      "basic tasks completed. We'll look at five fundamental things you might\n",
      "need to do:\n",
      "{ref}`process trees<sec_processing_trees>`,\n",
      "{ref}`sites & mutations<sec_processing_sites_and_mutations>`, and\n",
      "{ref}`genotypes<sec_processing_genotypes>`,\n",
      "{ref}`compute statistics<sec_tskit_getting_started_compute_statistics>`, and\n",
      "{ref}`save or export data<sec_tskit_getting_started_exporting_data>`.\n",
      "Throughout, we'll also provide pointers to where you can learn more.\n",
      "\n",
      ":::{note}\n",
      "The examples in this\n",
      "tutorial are all written using the {ref}`sec_python_api`, but it's also possible to\n",
      "{ref}`use R <sec_tskit_r>`, or access the API in other languages, notably\n",
      "{ref}`C<sec_c_api>` and [Rust](https://github.com/tskit-dev/tskit-rust).\n",
      ":::\n",
      "\n",
      "A number of different software programs can generate tree sequences. For the purposes\n",
      "of this tutorial we'll use [{program}`msprime`](https://tskit.dev/msprime) to create\n",
      "an example tree sequence representing the genetic genealogy of a 10Mb chromosome in\n",
      "twenty diploid individuals. To make it a bit more interesting, we'll simulate the effects\n",
      "of a {ref}`selective sweep<msprime:sec_ancestry_models_selective_sweeps>` in the middle\n",
      "of the chromosome, then throw some neutral mutations onto the resulting tree sequence.\n",
      "\n",
      "\n",
      "```{code-cell} ipython3\n",
      "import msprime\n",
      "\n",
      "pop_size=10_000\n",
      "seq_length=10_000_000\n",
      "\n",
      "sweep_model = msprime.SweepGenicSelection(\n",
      "    position=seq_length/2, start_frequency=0.0001, end_frequency=0.9999, s=0.25, dt=1e-6)\n",
      "\n",
      "ts = msprime.sim_ancestry(\n",
      "    20,\n",
      "    model=[sweep_model, msprime.StandardCoalescent()],\n",
      "    population_size=pop_size,\n",
      "    sequence_length=seq_length,\n",
      "    recombination_rate=1e-8,\n",
      "    random_seed=1234,  # only needed for repeatabilty\n",
      "    )\n",
      "# Optionally add finite-site mutations to the ts using the Jukes & Cantor model, creating SNPs\n",
      "ts = msprime.sim_mutations(ts, rate=1e-8, random_seed=4321)\n",
      "ts\n",
      "```\n",
      "\n",
      "You can see that there are many thousands of trees in this tree sequence.\n",
      "\n",
      ":::{note}\n",
      "Since we simulated the ancestry of 20 *diploid* individuals, our tree sequence\n",
      "contains 40 *sample nodes*, one for each genome.\n",
      ":::\n",
      "\n",
      "(sec_processing_trees)=\n",
      "\n",
      "## Processing trees\n",
      "\n",
      "Moving along a tree sequence usually involves iterating over all of its {class}`Tree`\n",
      "objects. This common idiom underlies many tree sequence algorithms, including those\n",
      "we'll encounter later in this tutorial for calculating \n",
      "{ref}`population genetic statistics<tskit:sec_stats>`.\n",
      "To iterate over a tree sequence you can use\n",
      "{meth}`TreeSequence.trees`.\n",
      "\n",
      "```{code-cell} ipython3\n",
      "for tree in ts.trees():\n",
      "    print(f\"Tree {tree.index} covers {tree.interval}\")\n",
      "    if tree.index >= 4:\n",
      "        print(\"...\")\n",
      "        break\n",
      "print(f\"Tree {ts.last().index} covers {ts.last().interval}\")\n",
      "```\n",
      "::::{margin}\n",
      ":::{caution}\n",
      "For efficiency, {meth}`~TreeSequence.trees` repeatedly returns the\n",
      "same tree object, updating it internally to reflect the (usually small) changes\n",
      "between adjacent trees. So take care to treat each tree within the ``for``\n",
      "loop separately, avoiding e.g. references outside the loop. The following\n",
      "produces a list of identical \"null\" trees:\n",
      "\n",
      "```\n",
      "# Don't do this!\n",
      "list(ts.trees())\n",
      "```\n",
      "\n",
      "If you need separate instances of each tree (inefficient, and will\n",
      "eat up your computer memory), use {meth}`~TreeSequence.aslist`.\n",
      ":::\n",
      "::::\n",
      "\n",
      "In this code snippet, as well as the {meth}`~TreeSequence.trees` iterator, we've also\n",
      "used {meth}`TreeSequence.last` to access the last tree directly; it may not surprise you\n",
      "that there's a corresponding {meth}`TreeSequence.first` method to return the first tree.\n",
      "The rest of the `tskit` tutorials will lead you through the concepts involved with\n",
      "storing and analysing sequences of many correlated trees. For a simple introduction, you\n",
      "might want to start with {ref}`sec_what_is`.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "--- ../../tskit/docs/popgen.md ---\n",
      "\n",
      "---\n",
      "jupytext:\n",
      "  text_representation:\n",
      "    extension: .md\n",
      "    format_name: myst\n",
      "    format_version: 0.13\n",
      "    jupytext_version: 1.10.3\n",
      "kernelspec:\n",
      "  display_name: Python 3\n",
      "  language: python\n",
      "  name: python3\n",
      "---\n",
      "\n",
      "```{currentmodule} tskit\n",
      "```\n",
      "\n",
      "```{code-cell}\n",
      ":tags: [remove-cell]\n",
      "import urllib.request\n",
      "\n",
      "import tqdm\n",
      "import tskit\n",
      "import tszip\n",
      "class TestLdOneSitePerTree(BaseTestLd):\n",
      "    @tests.cached_example\n",
      "    def ts(self):\n",
      "        ts = msprime.sim_ancestry(\n",
      "            5, sequence_length=10, recombination_rate=0.1, random_seed=1234\n",
      "        )\n",
      "        assert ts.num_trees > 3\n",
      "\n",
      "        tables = ts.dump_tables()\n",
      "        for tree in ts.trees():\n",
      "            site = tables.sites.add_row(tree.interval[0], ancestral_state=\"A\")\n",
      "            # Put the mutation somewhere deep in the tree\n",
      "            node = tree.preorder()[2]\n",
      "            tables.mutations.add_row(site=site, node=node, derived_state=\"B\")\n",
      "        ts = tables.tree_sequence()\n",
      "        # Return the full f2 matrix also\n",
      "        return ts, get_r2_matrix(ts)\n",
      "\n",
      "\n",
      "class TestLdAllSitesOneTree(BaseTestLd):\n",
      "    @tests.cached_example\n",
      "    def ts(self):\n",
      "        ts = msprime.sim_ancestry(\n",
      "            5, sequence_length=10, recombination_rate=0.1, random_seed=1234\n",
      "        )\n",
      "        assert ts.num_trees > 3\n",
      "\n",
      "        tables = ts.dump_tables()\n",
      "        tree = ts.at(5)\n",
      "        pos = np.linspace(tree.interval[0], tree.interval[1], num=10, endpoint=False)\n",
      "        for x, node in zip(pos, tree.preorder()[1:]):\n",
      "            site = tables.sites.add_row(x, ancestral_state=\"A\")\n",
      "            tables.mutations.add_row(site=site, node=node, derived_state=\"B\")\n",
      "        ts = tables.tree_sequence()\n",
      "        return ts, get_r2_matrix(ts)\n",
      "\n",
      "\n",
      "class TestLdSitesEveryOtherTree(BaseTestLd):\n",
      "    @tests.cached_example\n",
      "    def ts(self):\n",
      "        ts = msprime.sim_ancestry(\n",
      "            5, sequence_length=20, recombination_rate=0.1, random_seed=1234\n",
      "        )\n",
      "        assert ts.num_trees > 5\n",
      "\n",
      "        tables = ts.dump_tables()\n",
      "        for tree in ts.trees():\n",
      "            if tree.index % 2 == 0:\n",
      "                pos = np.linspace(*tree.interval, num=2, endpoint=False)\n",
      "                for x, node in zip(pos, tree.preorder()[1:]):\n",
      "                    site = tables.sites.add_row(x, ancestral_state=\"A\")\n",
      "                    tables.mutations.add_row(site=site, node=node, derived_state=\"B\")\n",
      "        ts = tables.tree_sequence()\n",
      "        return ts, get_r2_matrix(ts)\n",
      "\n",
      "\n",
      "class TestLdErrors:\n",
      "    def test_multi_mutations(self):\n",
      "        tables = tskit.TableCollection(2)\n",
      "        tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE, time=0)\n",
      "        tables.sites.add_row(position=0, ancestral_state=\"A\")\n",
      "        tables.sites.add_row(position=1, ancestral_state=\"A\")\n",
      "        tables.mutations.add_row(site=0, node=0, derived_state=\"C\")\n",
      "        tables.mutations.add_row(site=0, node=0, derived_state=\"T\", parent=0)\n",
      "        tables.mutations.add_row(site=1, node=0, derived_state=\"C\")\n",
      "        ts = tables.tree_sequence()\n",
      "        ldc = tskit.LdCalculator(ts)\n",
      "        with pytest.raises(tskit.LibraryError, match=\"Only infinite sites mutations\"):\n",
      "            ldc.r2(0, 1)\n",
      "        with pytest.raises(tskit.LibraryError, match=\"Only infinite sites mutations\"):\n",
      "            ldc.r2(1, 0)\n",
      "\n",
      "    @pytest.mark.parametrize(\"state\", [\"\", \"A\", \"AAAA\", \"💩\"])\n",
      "    def test_silent_mutations(self, state):\n",
      "        tables = tskit.TableCollection(2)\n",
      "        tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE, time=0)\n",
      "        tables.sites.add_row(position=0, ancestral_state=state)\n",
      "        tables.sites.add_row(position=1, ancestral_state=\"A\")\n",
      "        tables.mutations.add_row(site=0, node=0, derived_state=state)\n",
      "        tables.mutations.add_row(site=1, node=0, derived_state=\"C\")\n",
      "        ts = tables.tree_sequence()\n",
      "        ldc = tskit.LdCalculator(ts)\n",
      "        with pytest.raises(tskit.LibraryError, match=\"Silent mutations not supported\"):\n",
      "            ldc.r2(0, 1)\n",
      "        with pytest.raises(tskit.LibraryError, match=\"Silent mutations not supported\"):\n",
      "            ldc.r2(1, 0)\n",
      "class PythonTreeSequence:\n",
      "    \"\"\"\n",
      "    A python implementation of the TreeSequence object.\n",
      "\n",
      "    TODO this class is of limited use now and should be factored out as\n",
      "    part of a drive towards more modular versions of the tests currently\n",
      "    in tests_highlevel.py.\n",
      "    \"\"\"\n",
      "\n",
      "    def __init__(self, tree_sequence, breakpoints=None):\n",
      "        self._tree_sequence = tree_sequence\n",
      "        self._sites = []\n",
      "        # TODO this code here is expressed in terms of the low-level\n",
      "        # tree sequence for legacy reasons. It probably makes more sense\n",
      "        # to describe it in terms of the tables now if we want to have an\n",
      "        # independent implementation.\n",
      "        ll_ts = self._tree_sequence._ll_tree_sequence\n",
      "\n",
      "        def make_mutation(id_):\n",
      "            (\n",
      "                site,\n",
      "                node,\n",
      "                derived_state,\n",
      "                parent,\n",
      "                metadata,\n",
      "                time,\n",
      "                edge,\n",
      "            ) = ll_ts.get_mutation(id_)\n",
      "            return tskit.Mutation(\n",
      "                id=id_,\n",
      "                site=site,\n",
      "                node=node,\n",
      "                time=time,\n",
      "                derived_state=derived_state,\n",
      "                parent=parent,\n",
      "                metadata=metadata,\n",
      "                edge=edge,\n",
      "                metadata_decoder=tskit.metadata.parse_metadata_schema(\n",
      "                    ll_ts.get_table_metadata_schemas().mutation\n",
      "                ).decode_row,\n",
      "            )\n",
      "\n",
      "        for j in range(tree_sequence.num_sites):\n",
      "            pos, ancestral_state, ll_mutations, id_, metadata = ll_ts.get_site(j)\n",
      "            self._sites.append(\n",
      "                tskit.Site(\n",
      "                    id=id_,\n",
      "                    position=pos,\n",
      "                    ancestral_state=ancestral_state,\n",
      "                    mutations=[make_mutation(ll_mut) for ll_mut in ll_mutations],\n",
      "                    metadata=metadata,\n",
      "                    metadata_decoder=tskit.metadata.parse_metadata_schema(\n",
      "                        ll_ts.get_table_metadata_schemas().site\n",
      "                    ).decode_row,\n",
      "                )\n",
      "            )\n",
      "\n",
      "    def trees(self):\n",
      "        pt = PythonTree(self._tree_sequence.get_num_nodes())\n",
      "        pt.index = 0\n",
      "        for (left, right), rtt in tsutil.algorithm_R(self._tree_sequence):\n",
      "            pt.parent[:] = rtt.parent\n",
      "            pt.left_child[:] = rtt.left_child\n",
      "            pt.right_child[:] = rtt.right_child\n",
      "            pt.left_sib[:] = rtt.left_sib\n",
      "            pt.right_sib[:] = rtt.right_sib\n",
      "            pt.num_children[:] = rtt.num_children\n",
      "            pt.edge[:] = rtt.edge\n",
      "            pt.left_root = rtt.left_child[-1]\n",
      "            pt.left = left\n",
      "            pt.right = right\n",
      "            # Add in all the sites\n",
      "            pt.site_list = [\n",
      "                site for site in self._sites if left <= site.position < right\n",
      "            ]\n",
      "            yield pt\n",
      "            pt.index += 1\n",
      "        pt.index = -1\n",
      "class TestDeleteSites:\n",
      "    \"\"\"\n",
      "    Tests for the TreeSequence.delete_sites method\n",
      "    \"\"\"\n",
      "\n",
      "    def ts_with_4_sites(self):\n",
      "        ts = msprime.simulate(8, random_seed=3)\n",
      "        tables = ts.dump_tables()\n",
      "        tables.sites.set_columns(np.arange(0, 1, 0.25), *tskit.pack_strings([\"G\"] * 4))\n",
      "        tables.mutations.add_row(site=1, node=ts.first().parent(0), derived_state=\"C\")\n",
      "        tables.mutations.add_row(site=1, node=0, derived_state=\"T\", parent=0)\n",
      "        tables.mutations.add_row(site=2, node=1, derived_state=\"A\")\n",
      "        return tables.tree_sequence()\n",
      "\n",
      "    def test_remove_by_index(self):\n",
      "        ts = self.ts_with_4_sites().delete_sites([])\n",
      "        assert ts.num_sites == 4\n",
      "        assert ts.num_mutations == 3\n",
      "        ts = ts.delete_sites(2)\n",
      "        assert ts.num_sites == 3\n",
      "        assert ts.num_mutations == 2\n",
      "        ts = ts.delete_sites([1, 2])\n",
      "        assert ts.num_sites == 1\n",
      "        assert ts.num_mutations == 0\n",
      "\n",
      "    def test_remove_all(self):\n",
      "        ts = self.ts_with_4_sites().delete_sites(range(4))\n",
      "        assert ts.num_sites == 0\n",
      "        assert ts.num_mutations == 0\n",
      "        # should be OK to run on a siteless tree seq as no sites specified\n",
      "        ts.delete_sites([])\n",
      "\n",
      "    def test_remove_repeated_sites(self):\n",
      "        ts = self.ts_with_4_sites()\n",
      "        t1 = ts.delete_sites([0, 1], record_provenance=False)\n",
      "        t2 = ts.delete_sites([0, 0, 1], record_provenance=False)\n",
      "        t3 = ts.delete_sites([0, 0, 0, 1], record_provenance=False)\n",
      "        assert t1.tables == t2.tables\n",
      "        assert t1.tables == t3.tables\n",
      "\n",
      "    def test_remove_different_orders(self):\n",
      "        ts = self.ts_with_4_sites()\n",
      "        t1 = ts.delete_sites([0, 1, 3], record_provenance=False)\n",
      "        t2 = ts.delete_sites([0, 3, 1], record_provenance=False)\n",
      "        t3 = ts.delete_sites([3, 0, 1], record_provenance=False)\n",
      "        assert t1.tables == t2.tables\n",
      "        assert t1.tables == t3.tables\n",
      "\n",
      "    def test_remove_bad(self):\n",
      "        ts = self.ts_with_4_sites()\n",
      "        with pytest.raises(TypeError):\n",
      "            ts.delete_sites([\"1\"])\n",
      "        with pytest.raises(ValueError):\n",
      "            ts.delete_sites(4)\n",
      "        with pytest.raises(ValueError):\n",
      "            ts.delete_sites(-5)\n",
      "\n",
      "    def verify_removal(self, ts, remove_sites):\n",
      "        tables = ts.dump_tables()\n",
      "        tables.delete_sites(remove_sites)\n",
      "\n",
      "        # Make sure we've computed the mutation parents properly.\n",
      "        mutation_parent = tables.mutations.parent\n",
      "        tables.compute_mutation_parents()\n",
      "        assert np.array_equal(mutation_parent, tables.mutations.parent)\n",
      "\n",
      "        tsd = tables.tree_sequence()\n",
      "        assert tsd.num_sites == ts.num_sites - len(remove_sites)\n",
      "        source_sites = [site for site in ts.sites() if site.id not in remove_sites]\n",
      "        assert len(source_sites) == tsd.num_sites\n",
      "        for s1, s2 in zip(source_sites, tsd.sites()):\n",
      "            assert s1.position == s2.position\n",
      "            assert s1.ancestral_state == s2.ancestral_state\n",
      "            assert s1.metadata == s2.metadata\n",
      "            assert len(s1.mutations) == len(s2.mutations)\n",
      "            for m1, m2 in zip(s1.mutations, s2.mutations):\n",
      "                assert m1.node == m2.node\n",
      "                assert m1.derived_state == m2.derived_state\n",
      "                assert m1.metadata == m2.metadata\n",
      "\n",
      "        # Check we get the same genotype_matrix\n",
      "        G1 = ts.genotype_matrix()\n",
      "        G2 = tsd.genotype_matrix()\n",
      "        keep = np.ones(ts.num_sites, dtype=bool)\n",
      "        keep[remove_sites] = 0\n",
      "        assert np.array_equal(G1[keep], G2)\n",
      "\n",
      "    def test_simple_random_metadata(self):\n",
      "        ts = msprime.simulate(10, mutation_rate=10, random_seed=2)\n",
      "        ts = tsutil.add_random_metadata(ts)\n",
      "        assert ts.num_mutations > 5\n",
      "        self.verify_removal(ts, [1, 3])\n"
     ]
    }
   ],
   "source": [
    "output, context = generate(graph_state)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "```{currentmodule} tskit\n",
      "```\n",
      "\n",
      "(sec_tskit_getting_started)=\n",
      "# Getting started with {program}`tskit`\n",
      "\n",
      "You've run some simulations or inference methods, and you now have a \n",
      "{class}`TreeSequence` object; what now? This tutorial is aimed \n",
      "users who are new to {program}`tskit` and would like to get some\n",
      "basic tasks completed. We'll look at five fundamental things you might\n",
      "need to do:\n",
      "{ref}`process trees<sec_processing_trees>`,\n",
      "{ref}`sites & mutations<sec_processing_sites_and_mutations>`, and\n",
      "{ref}`genotypes<sec_processing_genotypes>`,\n",
      "{ref}`compute statistics<sec_tskit_getting_started_compute_statistics>`, and\n",
      "{ref}`save or export data<sec_tskit_getting_started_exporting_data>`.\n",
      "Throughout, we'll also provide pointers to where you can learn more.\n",
      "\n",
      ":::{note}\n",
      "The examples in this\n",
      "tutorial are all written using the {ref}`sec_python_api`, but it's also possible to\n",
      "{ref}`use R <sec_tskit_r>`, or access the API in other languages, notably\n",
      "{ref}`C<sec_c_api>` and [Rust](https://github.com/tskit-dev/tskit-rust).\n",
      ":::\n",
      "\n",
      "A number of different software programs can generate tree sequences. For the purposes\n",
      "of this tutorial we'll use [{program}`msprime`](https://tskit.dev/msprime) to create\n",
      "an example tree sequence representing the genetic genealogy of a 10Mb chromosome in\n",
      "twenty diploid individuals. To make it a bit more interesting, we'll simulate the effects\n",
      "of a {ref}`selective sweep<msprime:sec_ancestry_models_selective_sweeps>` in the middle\n",
      "of the chromosome, then throw some neutral mutations onto the resulting tree sequence.\n",
      "\n",
      "\n",
      "```{code-cell} ipython3\n",
      "import msprime\n",
      "\n",
      "pop_size=10_000\n",
      "seq_length=10_000_000\n",
      "\n",
      "sweep_model = msprime.SweepGenicSelection(\n",
      "    position=seq_length/2, start_frequency=0.0001, end_frequency=0.9999, s=0.25, dt=1e-6)\n",
      "\n",
      "ts = msprime.sim_ancestry(\n",
      "    20,\n",
      "    model=[sweep_model, msprime.StandardCoalescent()],\n",
      "    population_size=pop_size,\n",
      "    sequence_length=seq_length,\n",
      "    recombination_rate=1e-8,\n",
      "    random_seed=1234,  # only needed for repeatabilty\n",
      "    )\n",
      "# Optionally add finite-site mutations to the ts using the Jukes & Cantor model, creating SNPs\n",
      "ts = msprime.sim_mutations(ts, rate=1e-8, random_seed=4321)\n",
      "ts\n",
      "```\n",
      "\n",
      "You can see that there are many thousands of trees in this tree sequence.\n",
      "\n",
      ":::{note}\n",
      "Since we simulated the ancestry of 20 *diploid* individuals, our tree sequence\n",
      "contains 40 *sample nodes*, one for each genome.\n",
      ":::\n",
      "\n",
      "(sec_processing_trees)=\n",
      "\n",
      "## Processing trees\n",
      "\n",
      "Moving along a tree sequence usually involves iterating over all of its {class}`Tree`\n",
      "objects. This common idiom underlies many tree sequence algorithms, including those\n",
      "we'll encounter later in this tutorial for calculating \n",
      "{ref}`population genetic statistics<tskit:sec_stats>`.\n",
      "To iterate over a tree sequence you can use\n",
      "{meth}`TreeSequence.trees`.\n",
      "\n",
      "```{code-cell} ipython3\n",
      "for tree in ts.trees():\n",
      "    print(f\"Tree {tree.index} covers {tree.interval}\")\n",
      "    if tree.index >= 4:\n",
      "        print(\"...\")\n",
      "        break\n",
      "print(f\"Tree {ts.last().index} covers {ts.last().interval}\")\n",
      "```\n",
      "::::{margin}\n",
      ":::{caution}\n",
      "For efficiency, {meth}`~TreeSequence.trees` repeatedly returns the\n",
      "same tree object, updating it internally to reflect the (usually small) changes\n",
      "between adjacent trees. So take care to treat each tree within the ``for``\n",
      "loop separately, avoiding e.g. references outside the loop. The following\n",
      "produces a list of identical \"null\" trees:\n",
      "\n",
      "```\n",
      "# Don't do this!\n",
      "list(ts.trees())\n",
      "```\n",
      "\n",
      "If you need separate instances of each tree (inefficient, and will\n",
      "eat up your computer memory), use {meth}`~TreeSequence.aslist`.\n",
      ":::\n",
      "::::\n",
      "\n",
      "In this code snippet, as well as the {meth}`~TreeSequence.trees` iterator, we've also\n",
      "used {meth}`TreeSequence.last` to access the last tree directly; it may not surprise you\n",
      "that there's a corresponding {meth}`TreeSequence.first` method to return the first tree.\n"
     ]
    }
   ],
   "source": [
    "print(context[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from faiss_vector import repo_to_text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "repo_to_text(\"../../tskit/\", 'data-new.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "def create_notebook_data():\n",
      "    computing_statistics()\n",
      "    afs()\n",
      "\n",
      "# create_notebook_data()  # uncomment to recreate the tree seqs used in this notebook\n",
      "```\n",
      "\n",
      "(sec_analysing_tree_sequences)=\n",
      "\n",
      "# _Analysing tree sequences_\n",
      "% remove underscores in title when tutorial is complete or near-complete\n",
      "\n",
      ":::{note}\n",
      "This tutorial is a work in progress.\n",
      ":::\n",
      "\n",
      "\n",
      "(sec_tutorial_stats)=\n",
      "\n",
      "## Computing statistics\n",
      "\n",
      "Tskit provides an extensive and flexible interface for computing population\n",
      "genetic statistics, which is documented in detail in the\n",
      "{ref}`general statistics <sec_stats>` section of the offical documentation.\n",
      "This tutorial aims to give a quick overview of how the APIs work and how to use\n",
      "them effectively.\n",
      "\n",
      "First, let's load a tree sequence to work with which has roughly human\n",
      "parameters for 10 thousand samples and 10Mb chromosomes:\n",
      "\n",
      "```{code-cell} ipython3\n",
      "ts = tskit.load(\"data/computing_statistics.trees\")\n",
      "ts\n",
      "```\n",
      "\n",
      "This tree sequence has ~36.6 thousand trees & ~39 thousand segregating sites.\n",
      "We'd now like to compute some statistics on this dataset.\n",
      "\n",
      "### One-way statistics\n",
      "\n",
      "We refer to statistics that are defined with respect to a single set of\n",
      "samples as \"one-way\". An example of such a statistic is diversity, which\n",
      "is computed using the {meth}`TreeSequence.diversity` method:\n",
      "\n",
      "```{code-cell} ipython3\n",
      "d = ts.diversity()\n",
      "print(\"Average diversity per unit sequence length = {d:.3G}\")\n",
      "```\n",
      "\n",
      "This tells the average diversity across the whole sequence and returns a single\n",
      "number. We'll usually want to compute statistics in\n",
      "{ref}`windows <tskit:sec_stats_windows>` along the genome and we\n",
      "use the ``windows`` argument to do this:\n",
      "\n",
      "```{code-cell} ipython3\n",
      "windows = np.linspace(0, ts.sequence_length, num=5)\n",
      "d = ts.diversity(windows=windows)\n",
      "print(windows, d, sep=\"\\n\")\n",
      "```\n",
      "\n",
      "The ``windows`` argument takes a numpy array specifying the breakpoints\n",
      "along the genome. Here, we use numpy to create four equally spaced windows\n",
      "of size 2.5 megabases (the windows array contains k + 1 elements to define\n",
      "k windows). Because we have asked for values in windows, tskit now returns\n",
      "a numpy array rather than a single value. (See\n",
      "{ref}`sec_stats_output_dimensions` for a full description of how the output\n",
      "dimensions of statistics are determined by the ``windows`` argument.)\n",
      "\n",
      "Suppose we wanted to compute diversity within a specific subset of samples.\n",
      "We can do this using the ``sample_sets`` argument:\n",
      "\n",
      "```{code-cell} ipython3\n",
      "A = ts.samples()[:100]\n",
      "d = ts.diversity(sample_sets=A)\n",
      "print(d)\n",
      "```\n",
      "\n",
      "Here, we've computed the average diversity within the first hundred samples across\n",
      "the whole genome. As we've not specified any windows, this is again a single value.\n",
      "\n",
      "We can also compute diversity in *multiple* sample sets at the same time by providing\n",
      "a list of sample sets as an argument:\n",
      "\n",
      "```{code-cell} ipython3\n",
      "A = ts.samples()[:100]\n",
      "B = ts.samples()[100:200]\n",
      "C = ts.samples()[200:300]\n",
      "d = ts.diversity(sample_sets=[A, B, C])\n",
      "print(d)\n",
      "```\n",
      "\n",
      "Because we've computed multiple statistics concurrently, tskit returns a numpy array\n",
      "of these statistics. We have asked for diversity within three different sample sets,\n",
      "and tskit therefore returns an array with three values. (In general, the\n",
      "dimensions of the input determines the dimensions of the output: see\n",
      "{ref}`tskit:sec_stats_output_dimensions` for a detailed description of the rules.)\n",
      "\n",
      "We can also compute multiple statistics in multiple windows:\n",
      "\n",
      "```{code-cell} ipython3\n",
      "d = ts.diversity(sample_sets=[A, B, C], windows=windows)\n",
      "print(\"shape = \", d.shape, \"\\n\", d)\n",
      "```\n",
      "\n",
      "We have computed diversity within three different sample sets across four\n",
      "genomic windows, and our output is therefore a 2D numpy array with four\n",
      "rows and three columns: each row contains the diversity values within\n",
      "A, B and C for a particular window.\n",
      "\n",
      "### Multi-way statistics\n",
      "\n",
      "Many population genetic statistics compare multiple sets of samples to\n",
      "each other. For example, the {meth}`TreeSequence.divergence` method computes\n",
      "the divergence between two subsets of samples:\n"
     ]
    }
   ],
   "source": [
    "print(context[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dspy",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
